adapters/insticator/insticator.go

```diff
@@ -0,0 +1,328 @@
+			continue
+		}
+
+		// Populate site.publisher.id from imp extension
+		if requestCopy.Site != nil || requestCopy.App != nil {
+			if err := populatePublisherId(&impCopy, &requestCopy); err != nil {
+				errs = append(errs, err)
+				continue
+			}
+		}
+
+		resolvedBidFloor, errFloor := resolveBidFloor(impCopy.BidFloor, impCopy.BidFloorCur, requestInfo)
+		if errFloor != nil {
+			errs = append(errs, &errortypes.BadInput{
+				Message: fmt.Sprintf("Error in converting the provided bid floor currency from %s to USD",
+					impCopy.BidFloorCur),
+			})
+			continue
+		}
+		if resolvedBidFloor > 0 {
+			impCopy.BidFloor = resolvedBidFloor
+			impCopy.BidFloorCur = "USD"
+		}
+
+		groupedImps[impKey] = append(groupedImps[impKey], impCopy)
+	}
+
+	for _, impList := range groupedImps {
+		if adapterReq, err := a.makeRequest(&requestCopy, impList); err == nil {
+			adapterRequests = append(adapterRequests, adapterReq)
+		} else {
+			errs = append(errs, err)
+		}
+	}
+	return adapterRequests, errs
+}
+
+func (a *adapter) makeRequest(request *openrtb2.BidRequest, impList []openrtb2.Imp) (*adapters.RequestData, error) {
+	request.Imp = impList
+
+	reqJSON, err := jsonutil.Marshal(request)
+	if err != nil {
+		return nil, err
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+
+	if request.Device != nil {
+		if len(request.Device.UA) > 0 {
+			headers.Add("User-Agent", request.Device.UA)
+		}
+
+		if len(request.Device.IPv6) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IPv6)
+		}
+
+		if len(request.Device.IP) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IP)
+			headers.Add("IP", request.Device.IP)
+		}
+	}
+
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    reqJSON,
+		Headers: headers,
+		ImpIDs:  openrtb_ext.GetImpIDs(request.Imp),
+	}, nil
+}
+
+func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if adapters.IsResponseStatusCodeNoContent(responseData) {
+		return nil, nil
+	}
+
+	if err := adapters.CheckResponseStatusCodeForErrors(responseData); err != nil {
+		return nil, []error{err}
+	}
+
+	var response openrtb2.BidResponse
+	if err := jsonutil.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	if response.Cur != "" {
+		bidResponse.Currency = response.Cur
+	}
+	for _, seatBid := range response.SeatBid {
+		for i := range seatBid.Bid {
+			bid := &seatBid.Bid[i]
+			bidType := getMediaTypeForBid(bid)
+			b := &adapters.TypedBid{
+				Bid:     &seatBid.Bid[i],
+				BidType: bidType,
+			}
+			bidResponse.Bids = append(bidResponse.Bids, b)
+		}
+	}
+	return bidResponse, nil
+}
+
+func makeImps(imp openrtb2.Imp) (openrtb2.Imp, string, error) {
```

Comment:
It may make sense to return the publisher ID here, since you unmarshal the `imp.ext` here. This would save you from unmarshalling it again in `populatePublisherID()`. Or possibly move the unmarshal into the `Imp` loop and pass the unmarshalled structure into this function.

Reply:
I see @hhhjort 

what do you think of this ?
```
for i := 0; i < len(request.Imp); i++ {
                // I will return publisherId from the makeImps, and use it in next step
		impCopy, impKey, publisherId, err := makeImps(request.Imp[i])
		if err != nil {
			errs = append(errs, err)
			continue
		}

		// Populate site.publisher.id from imp extension
		if (requestCopy.Site != nil && (requestCopy.Site.Publisher == nil || requestCopy.Site.Publisher.ID == "")) ||
			(requestCopy.App != nil && (requestCopy.App.Publisher == nil || requestCopy.App.Publisher.ID == "")) {
			if err := populatePublisherId(publisherId, &requestCopy); err != nil {
				errs = append(errs, err)
				continue
			}
		}

... rest code
```